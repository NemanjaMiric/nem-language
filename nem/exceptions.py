"""Raise exceptions.

Holds multiple exceptions.

Exceptions:
LexerException - During lexing
ParserException - During parsing
EvaluatorException - During evaluation

"""


class LexerException(Exception):

    """Raise exception.

    Gets raised when an error appears during lexing.

    """

    pass


class ParserException(Exception):

    """Raise exception.

    Gets raised when an error appears during parsing.

    """

    pass


class EvaluatorException(Exception):

    """Raise exception.

    Gets raised when an error appears during evaluation.

    """

    pass