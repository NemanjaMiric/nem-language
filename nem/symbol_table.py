"""Hold a symbol table.

Holds class SymbolTable. Used for storing variable names and their values.

"""


class SymbolTable:

    """Store variable names and their values.

    Used for storing variable names and their values.

    """

    def __init__(self):
        """Initialize SymbolTable class."""
        self.parent = None
        self.symbols = {}

    def get(self, name):
        """Return the value of the symbol with the specified name."""
        temporary_value = self.symbols.get(name, None)

        if temporary_value is None and self.parent is not None:
            temporary_value = self.parent.get(name)

        return temporary_value

    def set(self, name, value):
        """Set the value of the symbol with the specified name."""
        self.symbols[name] = value
