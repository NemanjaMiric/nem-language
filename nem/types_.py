"""Represent type values.

Used for representing type values during evaluation.

"""


class Type:

    """Hold methods used by all types.

    Holds methods used by all types.

    """

    value = None

    def __bool__(self):
        """Convert self to bool."""
        return bool(self.value)

    def __or__(self, other):
        """Return value of OR operator when used with self and other class."""
        try:
            return Number(1 if self.value or other.value else 0)
        except TypeError:
            return Null()

    def __and__(self, other):
        """Return value of AND operator when used with self and other class."""
        try:
            return Number(1 if self.value and other.value else 0)
        except TypeError:
            return Null()

    def __eq__(self, other):
        """Return value of == operator when used with self and other class."""
        try:
            return Number(1 if self.value == other.value else 0)
        except TypeError:
            return Null()

    def __lt__(self, other):
        """Return value of < operator when used with self and other class."""
        try:
            return Number(1 if self.value < other.value else 0)
        except TypeError:
            return Null()

    def __le__(self, other):
        """Return value of <= operator when used with self and other class."""
        try:
            return Number(1 if self.value <= other.value else 0)
        except TypeError:
            return Null()

    def __gt__(self, other):
        """Return value of > operator when used with self and other class."""
        try:
            return Number(1 if self.value > other.value else 0)
        except TypeError:
            return Null()

    def __ge__(self, other):
        """Return value of >= operator when used with self and other class."""
        try:
            return Number(1 if self.value >= other.value else 0)
        except TypeError:
            return Null()


class Number(Type):

    """Hold type number.

    Holds type number.

    """

    def __init__(self, value):
        """Initialize Number class."""
        self.value = float(value)
        if self.value.is_integer():
            self.value = int(self.value)

    def __repr__(self):
        """Represent Number class."""
        return "Number({})".format(repr(self.value))

    def __str__(self):
        """Convert Number class to string."""
        return str(self.value)

    def __add__(self, other):
        """Add Number and other class."""
        if isinstance(other, Number):
            temporary = self.value + other.value
            return Number(temporary)
        raise NotImplementedError

    def __sub__(self, other):
        """Subtract Number and other class."""
        if isinstance(other, Number):
            temporary = self.value - other.value
            return Number(temporary)
        raise NotImplementedError

    def __mod__(self, other):
        """Get remainder of Number divided by other class."""
        if isinstance(other, Number):
            temporary = self.value % other.value
            return Number(temporary)
        raise NotImplementedError

    def __mul__(self, other):
        """Multiply Number and other class."""
        if isinstance(other, Number):
            temporary = self.value * other.value
            return Number(temporary)
        raise NotImplementedError

    def __truediv__(self, other):
        """Divide Number and other class."""
        if isinstance(other, Number):
            try:
                temporary = self.value / other.value
            except ZeroDivisionError:
                return Null()
            return Number(int(temporary) if temporary.is_integer() else temporary)
        raise NotImplementedError

    def __pow__(self, other):
        """Raise Number to the power of other class."""
        if isinstance(other, Number):
            temporary = self.value ** other.value
            return Number(temporary)
        raise NotImplementedError


class Text(Type):

    """Hold type text.

    Holds type text.

    """

    def __init__(self, value):
        """Initialize Text class."""
        self.value = value

    def __repr__(self):
        """Represent Text class."""
        return "Text({})".format(repr(self.value))

    def __str__(self):
        """Convert Text class to string."""
        return str(self.value)

    def __add__(self, other):
        """Concatenate Text with other class."""
        if isinstance(other, Text):
            return Text(self.value + other.value)
        raise NotImplementedError

    def __mul__(self, other):
        """Concatenate Text to itself a number of times."""
        if isinstance(other, Number):
            return Text(self.value * other.value)
        raise NotImplementedError


class List(Type):

    """Hold type list.

    Holds type list.

    """

    def __init__(self, value):
        """Initialize List class."""
        self.value = value

    def __repr__(self):
        """Represent List class."""
        return "List({})".format(repr(self.value))

    def __str__(self):
        """Convert Number class to string."""
        return "[{}]".format(", ".join(map(str, self.value)))

    def __add__(self, other):
        """Concatenate List with other class."""
        if isinstance(other, List):
            return List(self.value + other.value)
        raise NotImplementedError

    def __mul__(self, other):
        """Concatenate List to itself a number of times."""
        if isinstance(other, Number):
            return List(self.value * other.value)
        raise NotImplementedError


class Null(Type):

    """Hold type null.

    Holds type null.

    """

    def __repr__(self):
        """Represent Null class."""
        return "Null()"

    def __str__(self):
        """Convert Null class to string."""
        return "null"


class Function(Type):

    """Hold type function.

    Holds type function.

    """

    def __init__(self, parameters, body):
        """Initialize Function class."""
        self.parameters = parameters
        self.body = body

    def __repr__(self):
        """Represent Function class."""
        return "Function({}, {})".format(self.parameters, self.body)


class BuiltInFunction(Type):

    """Hold type built-in function.

    Holds type built-in function.

    """

    def __init__(self, parameters, index):
        """Initialize BuiltInFunction class."""
        self.parameters = parameters
        self.index = index

    def __repr__(self):
        """Represent Function class."""
        return "BuiltInFunction({}, {})".format(self.parameters, self.index)
