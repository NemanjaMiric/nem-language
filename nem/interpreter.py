"""Interpret Nem code.

Holds class Interpreter and used for debugging the class Interpreter.

"""


from nem.nemtoken import Token
from nem.exceptions import LexerException, ParserException
import re


class Interpreter:

    """Represent the interpreter.

    Used for representing the interpreter as a class.

    """

    def __init__(self, code):
        """Initialize the Interpreter class.

        Initializes the Interpreter class and adds a newline character to the end of the Nem code for handling EOF.

        """
        self.code = "{}\n".format(code)

    def lex(self):
        """Lex the code.

        Lexer for Nem code.

        List of tokens:
        ',' - <list-element>
        ';' - <end>
        r'[=+-*/^%<>]' - <operator>
        '(' - <left-bracket>
        ')' - <right-bracket>
        '[' - <left-square>
        ']' - <right-square>
        '"..."' - <text>
        r'[0-9]*[.]?[0-9]+' - <number>
        r'[a-z_][a-z_0-9]*' - <symbol>
        'null' - <null>
        'input' - <keyword>
        'output' - <keyword>
        'error' - <keyword>
        'while' - <keyword>
        'if' - <keyword>
        'otherwise' - <keyword>
        'import' - <keyword>
        'open' - <keyword>
        'convert' - <keyword>
        'number' - <keyword>
        'text' - <keyword>
        'list' - <keyword>
        'break' - <keyword>
        'continue' - <keyword>
        'function' - <keyword>
        'not' - <operator>
        'is' - <operator>
        'and' - <operator>
        'or' - <operator>

        Gets ignored:
        '#...\n' - Comment

        """
        tokens = []
        current_index = 0

        line = 1
        character = 1

        while current_index < len(self.code):
            current_token = self.code[current_index]

            if current_token == "\n":
                line += 1
                character = 1

            elif current_token == ",":
                tokens.append(Token("list-element", current_token))

            elif current_token == ";":
                tokens.append(Token("end", current_token))

            elif current_token in "=+-*/^%<>":
                tokens.append(Token("operator", current_token))

            elif current_token == "[":
                tokens.append(Token("left-square", current_token))

            elif current_token == "]":
                tokens.append(Token("right-square", current_token))

            elif current_token == "(":
                tokens.append(Token("left-bracket", current_token))

            elif current_token == ")":
                tokens.append(Token("right-bracket", current_token))

            elif current_token == "#":
                while self.code[current_index] != "\n":
                    current_index += 1
                line += 1
                character = 0

            elif current_token == "\"":
                current_index += 1
                start_index = current_index
                try:
                    while self.code[current_index] != "\"":
                        current_index += 1
                        character += 1
                except IndexError:
                    raise LexerException("Lexing Error (Line {}, Character: {}): Text doesn't have an end"
                                         .format(line, character))
                else:
                    tokens.append(Token("text", self.code[start_index:current_index]))

            elif re.match("[a-zA-Z_]", current_token):
                start_index = current_index
                current_index += 1
                character += 1
                while re.match("[a-zA-Z_][a-zA-Z_0-9]*\Z", self.code[start_index:current_index + 1]):
                    current_index += 1
                    character += 1
                if self.code[start_index:current_index] in ("input", "output", "error", "while", "if",
                                                            "otherwise", "import", "open", "convert", "number",
                                                            "text", "list", "break", "continue", "function"
                                                            ):
                    tokens.append(Token("keyword", self.code[start_index:current_index]))
                elif self.code[start_index:current_index] == "null":
                    tokens.append(Token("null", self.code[start_index:current_index]))
                elif self.code[start_index:current_index] in ("not", "is", "and", "or"):
                    tokens.append(Token("operator", self.code[start_index:current_index]))
                else:
                    tokens.append(Token("symbol", self.code[start_index:current_index]))
                continue

            elif re.match("[0-9.]", current_token):
                start_index = current_index
                current_index += 1
                character += 1
                while re.match("[0-9]*[.]?[0-9]+\Z", self.code[start_index:current_index + 1])                         \
                        or self.code[start_index:current_index + 1] == ".":
                    current_index += 1
                    character += 1
                if self.code[start_index:current_index] == ".":
                    raise LexerException("Lexing Error (Line {}, Character: {}): Unrecognized token"
                                         .format(line, character))
                else:
                    tokens.append(Token("number", self.code[start_index:current_index]))
                continue

            elif current_token in " \t":
                pass

            else:
                raise LexerException("Lexing Error (Line {}, Character: {}): Unrecognized token"
                                     .format(line, character))

            current_index += 1
            character += 1

        return tokens

    def parse(self):
        """Parse the tokens.

        Parses the tokens that get returned from the lexer and builds an AST out of them.

        """
        pass


def main():
    """Debug the interpreter.

    Used as the entry-point when the file gets ran directly. It's usually used for debugging the class Interpreter.

    """
    # For Debug purposes
    while True:
        print(Interpreter(input(">> ")).lex())


if __name__ == "__main__":
    # Only activates when the file gets ran directly.
    main()
