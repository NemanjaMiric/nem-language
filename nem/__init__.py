"""Initialize modules."""

import nem.evaluator
import nem.exceptions
import nem.interpreter
import nem.lexer
import nem.nodes
import nem.parser
import nem.symbol_table
import nem.token_
import nem.types_
