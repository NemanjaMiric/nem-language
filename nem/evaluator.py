"""Hold Evaluator class.

Holds the Evaluator class which is used for evaluating the Abstract Syntax Tree created by the parser.

"""

from nem.exceptions import EvaluatorException
import nem.interpreter
import nem.nodes as ast
import nem.types_ as value


class Evaluator:

    """Evaluate the AST.

    Used for evaluating the Abstract Syntax Tree created by the parser.

    """

    def __init__(self, abstract_syntax_tree, symbol_table):
        """Initialize Evaluator class."""
        self.abstract_syntax_tree = abstract_syntax_tree
        self.symbol_table = symbol_table

    @staticmethod
    def _check(value_, return_error=None, continue_error=None, break_error=None):
        """Check for return, continue and break."""
        try:
            if value_[0] == "RETURN":
                if return_error is None:
                    return value_, True
                else:
                    raise EvaluatorException(return_error)
            elif value_[0] == "CONTINUE":
                if continue_error is None:
                    return value_, True
                else:
                    raise EvaluatorException(continue_error)
            elif value_[0] == "BREAK":
                if continue_error is None:
                    return value_, True
                else:
                    raise EvaluatorException(break_error)
        except TypeError:
            pass

        return value_, False

    @staticmethod
    def _evaluate_number(node, _):
        """Evaluate Number node."""
        return value.Number(node.value)

    @staticmethod
    def _evaluate_variable(node, symbol_table):
        """Evaluate Variable node."""
        temporary_variable = symbol_table.get(node.variable)

        if temporary_variable is not None:
            return temporary_variable
        else:
            raise EvaluatorException("Evaluation Error (File {}) (Line {}): Variable '{}' is not defined"
                                     .format(node.filename, node.line, node.variable))

    @staticmethod
    def _evaluate_text(node, _):
        """Evaluate Text node."""
        return value.Text(node.text)

    def _evaluate_list(self, node, symbol_table):
        """Evaluate List node."""
        temporary_elements = []

        for temporary_element in node.elements:
            temporary_value = self._check(
                self._evaluate_node(temporary_element, symbol_table),

                "Evaluation Error (File {}) (Line {}): Cannot return value inside of list"
                .format(node.filename, node.line),
                "Evaluation Error (File {}) (Line {}): Cannot continue inside of list"
                .format(node.filename, node.line),
                "Evaluation Error (File {}) (Line {}): Cannot break inside of list"
                .format(node.filename, node.line)
            )[0]

            temporary_elements.append(temporary_value)

        return value.List(temporary_elements)

    def _evaluate_listindex(self, node, symbol_table):
        """Evaluate ListIndex node."""
        temporary_holder = self._evaluate_node(node.holder, symbol_table)

        temporary_index = self._check(
            self._evaluate_node(node.index, symbol_table),

            "Evaluation Error (File {}) (Line {}): Cannot return value inside of list index"
            .format(node.filename, node.line),
            "Evaluation Error (File {}) (Line {}): Cannot continue inside of list index"
            .format(node.filename, node.line),
            "Evaluation Error (File {}) (Line {}): Cannot break inside of list index"
            .format(node.filename, node.line)
        )[0]

        if not isinstance(temporary_index, value.Number):
            raise EvaluatorException(
                "Evaluation Error (File {}) (Line {}): List index has to be a number"
                .format(node.filename, node.line)
            )

        try:
            # Text indexing
            if isinstance(temporary_holder, value.Text):
                return value.Text(temporary_holder.value[temporary_index.value])
            # List indexing
            elif isinstance(temporary_holder, value.List):
                return temporary_holder.value[temporary_index.value]
            else:
                raise EvaluatorException(
                    "Evaluation Error (Filename {}) (Line {}): '{}' does not support indexing"
                    .format(node.filename, node.line, type(temporary_holder).__name__)
                )
        except IndexError:
            raise EvaluatorException(
                "Evaluation Error (Filename {}) (Line {}): Index out of range"
                .format(node.filename, node.line)
            )

    def _evaluate_expressions(self, node, symbol_table):
        """Evaluate Expressions node."""
        temporary_value = (value.Null(), False)

        for expression in node.expressions:
            temporary_value = self._check(self._evaluate_node(expression, symbol_table))

            # If 'return', 'continue' or 'break' is detected, backpropagate
            if temporary_value[1]:
                return temporary_value[0]

        return temporary_value[0]

    def _evaluate_unaryoperation(self, node, symbol_table):
        """Evaluate UnaryOperation node."""
        temporary_value = self._check(
            self._evaluate_node(node.node, symbol_table),

            "Evaluation Error (File {}) (Line {}): Cannot perform unary operation on return"
            .format(node.filename, node.line),
            "Evaluation Error (File {}) (Line {}): Cannot perform unary operation on continue"
            .format(node.filename, node.line),
            "Evaluation Error (File {}) (Line {}): Cannot perform unary operation on break"
            .format(node.filename, node.line)
        )[0]

        if node.operator == "not":
            if isinstance(temporary_value, value.Number):
                return value.Number(0 if temporary_value.value else 1)
            elif isinstance(temporary_value, value.Text):
                return value.Number(0 if temporary_value.value else 1)
            elif isinstance(temporary_value, value.List):
                return value.Number(0 if temporary_value.value else 1)
            elif isinstance(temporary_value, value.Null):
                return value.Number(1)
            else:
                raise EvaluatorException(
                    "Evaluation Error (File {}) (Line {}): Cannot apply unary operator 'not' on '{}'"
                    .format(node.filename, node.line, type(temporary_value).__name__)
                )
        elif node.operator == "+":
            if isinstance(temporary_value, value.Number):
                return temporary_value
            else:
                raise EvaluatorException(
                    "Evaluation Error (File {}) (Line {}): Cannot apply unary operator '+' on '{}'"
                    .format(node.filename, node.line, type(temporary_value).__name__)
                )
        elif node.operator == "-":
            if isinstance(temporary_value, value.Number):
                return value.Number(-temporary_value.value)
            else:
                raise EvaluatorException(
                    "Evaluation Error (File {}) (Line {}): Cannot apply unary operator '-' on '{}'"
                    .format(node.filename, node.line, type(temporary_value).__name__)
                )
        else:
            raise EvaluatorException(
                "Evaluation Error (File {}) (Line {}): Unary operator '{}' not implemented"
                .format(node.filename, node.line, node.operator)
            )

    def _evaluate_binaryoperation(self, node, symbol_table):
        """Evaluate BinaryOperation node."""
        temporary_left_value = self._check(self._evaluate_node(node.left_node, symbol_table))

        # If 'return', 'continue' or 'break' is detected, backpropagate
        if temporary_left_value[1]:
            return temporary_left_value[0]
        temporary_left_value = temporary_left_value[0]

        temporary_right_value = self._check(self._evaluate_node(node.right_node, symbol_table))

        # If 'return', 'continue' or 'break' is detected, backpropagate
        if temporary_right_value[1]:
            return temporary_right_value[0]
        temporary_right_value = temporary_right_value[0]

        try:
            if node.operator == "+":
                return temporary_left_value + temporary_right_value
            elif node.operator == "-":
                return temporary_left_value - temporary_right_value
            elif node.operator == "%":
                return temporary_left_value % temporary_right_value
            elif node.operator == "*":
                return temporary_left_value * temporary_right_value
            elif node.operator == "/":
                return temporary_left_value / temporary_right_value
            elif node.operator == "^":
                return temporary_left_value ** temporary_right_value
            elif node.operator == "or":
                return temporary_left_value or temporary_right_value
            elif node.operator == "and":
                return temporary_left_value and temporary_right_value
            elif node.operator == "is":
                return temporary_left_value == temporary_right_value
            elif node.operator == "<":
                return temporary_left_value < temporary_right_value
            elif node.operator == "<=":
                return temporary_left_value <= temporary_right_value
            elif node.operator == ">":
                return temporary_left_value > temporary_right_value
            elif node.operator == ">=":
                return temporary_left_value >= temporary_right_value
            else:
                raise EvaluatorException(
                    "Evaluation Error (File {}) (Line {}): Binary operator '{}' not implemented"
                    .format(node.filename, node.line, node.operator)
                )
        except (NotImplementedError, TypeError):
            raise EvaluatorException(
                "Evaluation Error (File {}) (Line {}): Cannot apply binary operator '{}' to '{}' and '{}'"
                .format(
                    node.filename, node.line, node.operator,
                    type(temporary_left_value).__name__, type(temporary_right_value).__name__
                )
            )

    def _evaluate_assignmentoperation(self, node, symbol_table):
        """Evaluate AssignmentOperation node."""
        temporary_value = self._check(
            self._evaluate_node(node.value, symbol_table),

            "Evaluation Error (File {}) (Line {}): Cannot assign variable to return value"
            .format(node.filename, node.line),
            "Evaluation Error (File {}) (Line {}): Cannot assign variable to continue"
            .format(node.filename, node.line),
            "Evaluation Error (File {}) (Line {}): Cannot assign variable to break"
            .format(node.filename, node.line)
        )[0]

        symbol_table.set(node.variable, temporary_value)

        return temporary_value

    def _evaluate_ifotherwise(self, node, symbol_table):
        """Evaluate IfOtherwise node."""
        temporary_condition = self._check(
            self._evaluate_node(node.condition, symbol_table),

            "Evaluation Error (File {}) (Line {}): Cannot return value in condition"
            .format(node.filename, node.line),
            "Evaluation Error (File {}) (Line {}): Cannot continue in condition"
            .format(node.filename, node.line),
            "Evaluation Error (File {}) (Line {}): Cannot break in condition"
            .format(node.filename, node.line)
        )[0]

        # If condition is true
        if temporary_condition.value:
            temporary_expression = self._check(self._evaluate_node(node.if_expression, symbol_table))

            # If 'return', 'continue' or 'break' is detected, backpropagate
            if temporary_expression[1]:
                return temporary_expression[0]
        # Otherwise
        elif node.otherwise_expression is not None:
            temporary_expression = self._check(self._evaluate_node(node.otherwise_expression, symbol_table))

            # If 'return', 'continue' or 'break' is detected, backpropagate
            if temporary_expression[1]:
                return temporary_expression[0]

        return value.Null()

    def _evaluate_while(self, node, symbol_table):
        """Evaluate While node."""
        temporary_condition = self._check(
            self._evaluate_node(node.condition, symbol_table),

            "Evaluation Error (File {}) (Line {}): Cannot return value in condition"
            .format(node.filename, node.line),
            "Evaluation Error (File {}) (Line {}): Cannot continue in condition"
            .format(node.filename, node.line),
            "Evaluation Error (File {}) (Line {}): Cannot break in condition"
            .format(node.filename, node.line)
        )[0]

        while temporary_condition.value:
            temporary_return_value = self._check(self._evaluate_node(node.expression, symbol_table))

            # If 'break' is detected, break the loop
            if temporary_return_value[1] and temporary_return_value[0][0] == "BREAK":
                break

            temporary_condition = self._evaluate_node(node.condition, symbol_table)

        return value.Null()

    @staticmethod
    def _evaluate_functiondefinition(node, symbol_table):
        """Evaluate FunctionDefinition node."""
        temporary_function = value.Function(node.parameters, node.expression)
        symbol_table.set(node.name, temporary_function)
        return temporary_function

    def _evaluate_functioncall(self, node, symbol_table):
        """Evaluate FunctionCall node."""
        temporary_function = symbol_table.get(node.name)

        if temporary_function is None:
            if isinstance(node.name, ast.FunctionCall):
                temporary_function = self._evaluate_functioncall(node.name, symbol_table)
            else:
                raise EvaluatorException(
                    "Evaluation Error (File {}) (Line {}): Function '{}' is not defined"
                    .format(node.filename, node.line, node.name)
                )

        difference = len(temporary_function.parameters) - len(node.arguments)

        if difference > 0:
            raise EvaluatorException(
                "Evaluation Error (File {}) (Line {}): Function call for function '{}' is missing {} argument(s)"
                .format(node.filename, node.line, node.name, difference)
            )
        elif difference < 0:
            raise EvaluatorException(
                "Evaluation Error (File {}) (Line {}): Function call for function '{}' has too many argument(s) ({})"
                .format(node.filename, node.line, node.name, abs(difference))
            )

        temporary_symbol_table = symbol_table.copy()

        for parameter, argument in zip(temporary_function.parameters, node.arguments):
            argument = self._check(
                self._evaluate_node(argument, temporary_symbol_table),

                "Evaluation Error (File {}) (Line {}): Cannot return value in argument"
                .format(node.filename, node.line),
                "Evaluation Error (File {}) (Line {}): Cannot continue in argument"
                .format(node.filename, node.line),
                "Evaluation Error (File {}) (Line {}): Cannot break in argument"
                .format(node.filename, node.line)
            )[0]

            temporary_symbol_table.set(parameter, argument)

        if isinstance(temporary_function, value.Function):
            temporary_return_value = self._check(self._evaluate_node(temporary_function.body, temporary_symbol_table))

            # If 'return' is detected, return the value
            if temporary_return_value[1] and temporary_return_value[0][0] == "RETURN":
                return temporary_return_value[0][1]

            return value.Null()
        elif isinstance(temporary_function, value.BuiltInFunction):
            # 'print(element)' built-in function
            if temporary_function.index == 0:
                parameter_element = temporary_symbol_table.get("element")

                print(parameter_element, end="")

                return value.Null()
            # 'input()' built-in function
            elif temporary_function.index == 1:
                return value.Text(input())
            # 'convert(value, type)' built-in function
            elif temporary_function.index == 2:
                parameter_value = temporary_symbol_table.get("value")
                parameter_type = temporary_symbol_table.get("type")

                # Convert Number to...
                if isinstance(parameter_value, value.Number):
                    # ... Number
                    if parameter_type.value == "number":
                        return parameter_value
                    # ... Text
                    elif parameter_type.value == "text":
                        return value.Text(str(parameter_value.value))
                    # ... List
                    elif parameter_type.value == "list":
                        return value.List(list(str(parameter_value.value)))
                    else:
                        return parameter_value
                # Convert Text to...
                elif isinstance(parameter_value, value.Text):
                    # ... Number
                    if parameter_type.value == "number":
                        try:
                            return value.Number(float(parameter_value.value))
                        except ValueError:
                            return parameter_value
                    # ... Text
                    elif parameter_type.value == "text":
                        return parameter_value
                    # ... List
                    elif parameter_type.value == "list":
                        return value.List(list(str(parameter_value.value)))
                    else:
                        return parameter_value
                # Convert List to...
                elif isinstance(parameter_value, value.List):
                    # ... Number
                    if parameter_type.value == "number":
                        try:
                            return value.Number(float("".join(map(str, parameter_value.value))))
                        except ValueError:
                            return parameter_value
                    # ... Text
                    elif parameter_type.value == "text":
                        return value.Text("".join(map(str, parameter_value.value)))
                    # ... List
                    elif parameter_type.value == "list":
                        return parameter_value
                    else:
                        return parameter_value
                else:
                    return parameter_value
            else:
                raise EvaluatorException(
                    "Evaluation Error (File {}) (Line {}): Built-in function '{}' not implemented"
                    .format(node.filename, node.line, node.name)
                )
        else:
            raise EvaluatorException(
                "Evaluation Error (File {}) (Line {}): Function type '{}' not implemented"
                .format(node.filename, node.line, type(temporary_function))
            )

    @staticmethod
    def _evaluate_import(node, symbol_table):
        """Evaluate Import node."""
        try:
            with open("{}.nem".format(node.file)) as nem_file:
                temporary_code = nem_file.read()
                nem.interpreter.Interpreter(temporary_code, node.file, symbol_table)
        except FileNotFoundError:
            raise EvaluatorException(
                "Evaluation Error (File {}) (Line {}): File '{}.nem' doesn't exist"
                .format(node.filename, node.line, node.file)
            )

        return value.Null()

    def _evaluate_return(self, node, symbol_table):
        """Evaluate Return node."""
        temporary_return_value = self._check(
            self._evaluate_node(node.value, symbol_table),

            "Evaluation Error (File {}) (Line {}): Cannot return value within return value"
            .format(node.filename, node.line),
            "Evaluation Error (File {}) (Line {}): Cannot continue within return value"
            .format(node.filename, node.line),
            "Evaluation Error (File {}) (Line {}): Cannot break within return value"
            .format(node.filename, node.line)
        )[0]

        return "RETURN", temporary_return_value

    @staticmethod
    def _evaluate_continue(_, __):
        """Evaluate Continue node."""
        return "CONTINUE", None

    @staticmethod
    def _evaluate_break(_, __):
        """Evaluate Break node."""
        return "BREAK", None

    @staticmethod
    def _evaluate_null(_, __):
        """Evaluate Null node."""
        return value.Null()

    def _evaluate_node(self, node, symbol_table):
        """Evaluate a node."""
        try:
            method = getattr(self, "_evaluate_{}".format(type(node).__name__.lower()))
        except AttributeError:
            raise EvaluatorException(
                "Evaluation Error (File {}) (Line {}): Node '{}' is not defined"
                .format(node.filename, node.line, type(node).__name__)
            )

        return method(node, symbol_table)

    def _evaluate_all(self):
        """Evaluate all nodes."""
        for node in self.abstract_syntax_tree:
            current_result = self._check(
                self._evaluate_node(node, self.symbol_table),

                "Evaluation Error (File {}) (Line {}): 'return' not in function"
                .format(node.filename, node.line),
                "Evaluation Error (File {}) (Line {}): 'continue' not in loop"
                .format(node.filename, node.line),
                "Evaluation Error (File {}) (Line {}): 'break' not in loop"
                .format(node.filename, node.line)
            )[0]

            yield current_result

    def evaluate(self):
        """Evaluate the AST.

        Evaluator for Nem code. Evaluates the Abstract Syntax Tree created by the parser.

        """
        return self._evaluate_all()


def main():
    """Debug the evaluator.

    Used as the entry-point when the file gets ran directly. Used for debugging the class Evaluator.

    """
    symbol_table = SymbolTable()
    symbol_table.set("true", value.Number(1))
    symbol_table.set("false", value.Number(0))
    symbol_table.set("print", value.BuiltInFunction(["element"], 0))
    symbol_table.set("input", value.BuiltInFunction([], 1))
    symbol_table.set("convert", value.BuiltInFunction(["value", "type"], 2))

    while True:
        print(list(Evaluator(Parser(Lexer(input(">> ") + "\n", "<stdin>").lex()).parse(), symbol_table).evaluate()))


if __name__ == "__main__":
    # Only activates when the file gets ran directly.
    from nem.lexer import Lexer
    from nem.parser import Parser
    from nem.symbol_table import SymbolTable

    main()
