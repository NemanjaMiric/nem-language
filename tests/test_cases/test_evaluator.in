# Fibonacci numbers generator

function fib(first, second, stop) (
    while (not 0 and not 0) (
        if(first < stop) (
            temporary = first
            first = second
            second = temporary + second
        ) otherwise break
        continue
    )

    return 0
)

fib(1, 1, 1000)

a1 = [1, null, "\"text\""]
_0 = (- 0.2 + .2) ^ (2 % 2) * (-2 + 02) ^ (2 - 2) + (-2 - 2) ^ (2 + 2) / (2 + 2) ^ (2 + 2) # comments are cool

b = a1[1]

function a() (
    function b()
        return [1, [1, 2, [2, 3]]]

    return b
)

d = convert(a()()[1][2][0], "number")
print([1, 2][0])
print("asd"[1][0])
