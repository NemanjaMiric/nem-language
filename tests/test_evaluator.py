"""Test for Evaluator class.

Unit testing for the Evaluator class.

"""


import unittest
from nem.lexer import Lexer
from nem.parser import Parser
from nem.evaluator import Evaluator
from nem.symbol_table import SymbolTable
from nem.types_ import *
from nem.exceptions import EvaluatorException


class ParserTestCase(unittest.TestCase):

    """Unit test Parser class.

    Used for unit testing the Parser class.

    """

    def test_parse(self):
        """Test the parse method.

        Tests the parser.

        """
        symbol_table = SymbolTable()
        symbol_table.set("true", Number(1))
        symbol_table.set("false", Number(0))
        symbol_table.set("print", BuiltInFunction(["element"], 0))
        symbol_table.set("input", BuiltInFunction([], 1))
        symbol_table.set("convert", BuiltInFunction(["value", "type"], 2))

        with open("test_cases/test_evaluator.in") as _input:
            _input = _input.read()

        with open("test_cases/test_evaluator.out") as output:
            output = output.read()

        self.assertEqual(
            "".join(list(map(repr, Evaluator(Parser(Lexer(_input, "test").lex()).parse(), symbol_table).evaluate()))),
            output
        )

        with self.assertRaises(EvaluatorException):
            list(Evaluator(Parser(Lexer("\nvariable", "<stdin>").lex()).parse(), symbol_table).evaluate())

        with self.assertRaises(EvaluatorException):
            list(Evaluator(Parser(Lexer("import \"test\"\n", "<stdin>").lex()).parse(), symbol_table).evaluate())

        with self.assertRaises(EvaluatorException):
            list(Evaluator(Parser(Lexer("nem(1, 2)\n", "<stdin>").lex()).parse(), symbol_table).evaluate())

        with self.assertRaises(EvaluatorException):
            list(Evaluator(Parser(Lexer("[1, 2][2]\n", "<stdin>").lex()).parse(), symbol_table).evaluate())

        with self.assertRaises(EvaluatorException):
            list(Evaluator(Parser(Lexer("22 * [1, 2]\n", "<stdin>").lex()).parse(), symbol_table).evaluate())

        with self.assertRaises(EvaluatorException):
            list(Evaluator(Parser(Lexer("[1, 2] ^ [1, 2]\n", "<stdin>").lex()).parse(), symbol_table).evaluate())

        with self.assertRaises(EvaluatorException):
            list(Evaluator(Parser(Lexer("a = return 1\n", "<stdin>").lex()).parse(), symbol_table).evaluate())

        with self.assertRaises(EvaluatorException):
            list(Evaluator(Parser(Lexer("while (return 1) (print(1))\n", "<stdin>").lex()).parse(),
                           symbol_table).evaluate())

        with self.assertRaises(EvaluatorException):
            list(Evaluator(Parser(Lexer("if (return 1) (print(1))\n", "test").lex()).parse(), symbol_table).evaluate())

        with self.assertRaises(EvaluatorException):
            list(Evaluator(Parser(Lexer("return 2\n", "<stdin>").lex()).parse(), symbol_table).evaluate())


if __name__ == '__main__':
    # Only activates when the file gets ran directly.
    unittest.main()
