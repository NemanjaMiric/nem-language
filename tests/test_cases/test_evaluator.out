Function(['first', 'second', 'stop'], Expressions([While(Expressions([BinaryOperation(UnaryOperation(Number('0'), 'not'), 'and', UnaryOperation(Number('0'), 'not'))]), Expressions([IfOtherwise(Expressions([BinaryOperation(Variable('first'), '<', Variable('stop'))]), Expressions([AssignmentOperation('temporary', Variable('first')), AssignmentOperation('first', Variable('second')), AssignmentOperation('second', BinaryOperation(Variable('temporary'), '+', Variable('second')))]), Break()), Continue()])), Return(Number('0'))]))Number(0)List([Number(1), Null(), Text('"text"')])Number(2)Null()Function([], Expressions([FunctionDefinition('b', [], Return(List([Number('1'), List([Number('1'), Number('2'), List([Number('2'), Number('3')])])]))), Return(Variable('b'))]))Number(2)Null()Null()